flink-yarn/src/test/java/org/apache/flink/yarn/YarnFileStageTest.java

```diff
@@ -126,7 +130,18 @@ public void testCopyFromLocalRecursiveWithoutScheme() throws Exception {
 		final FileSystem targetFileSystem = hdfsRootPath.getFileSystem(hadoopConfig);
 		final Path targetDir = targetFileSystem.getWorkingDirectory();
 
-		testCopyFromLocalRecursive(targetFileSystem, targetDir, tempFolder, false);
+		testCopyFromLocalRecursive(targetFileSystem, targetDir, LOCAL_RESOURCE_DIRECTORY, tempFolder, false);
+	}
+
+	/**
+	 * Verifies that a single file is properly copied.
+	 */
+	@Test
+	public void testCopyAFileFromLocal() throws IOException, URISyntaxException, InterruptedException {
```

Comment:
I currently don't see how this test adds additional coverage. If we can upload a directory recursively, we can also assume that uploading a single file in a directory works.

Reply:
The `uploadAndRegisterFiles` support to upload both a directory and a file.
I add this test for verifying uploading a file to the `usrlib` directory correctly. This is a different code path with uploading files in a directory.

Reply:
Alright, I can see now there is a different branch. However, there is quite some redundancy between `testCopyAFileFromLocal` and `testCopyFromLocalRecursive`. Unfortunately, I cannot give you a solution on the spot. BTW, your test would will pass even if we simplify the code in `uploadAndRegisterFiles` to

```
	for (File shipFile : shipFiles) {
			// add directories to the classpath
			final java.nio.file.Path shipPath = shipFile.toPath();
			final java.nio.file.Path parentPath = shipPath.getParent();
			Files.walkFileTree(shipPath, new SimpleFileVisitor<java.nio.file.Path>() {
				@Override
				public FileVisitResult visitFile(java.nio.file.Path file, BasicFileAttributes attrs) {
					localPaths.add(new Path(file.toUri()));
					relativePaths.add(new Path(localResourcesDirectory, parentPath.relativize(file).toString()));
					return FileVisitResult.CONTINUE;
				}
			});
	}
``` 

The reason is that `Files.walkFileTree` also works if the first argument is not a directory.

Reply:
If the `shipFile` is a file which is in a relative style for example "a.jar" `shipPath.getParent()` would be null.

I could make the`shipFiles` to the absolute path to avoid this situation and remove the new test.
I think this could work after some investigations. However, because this branch is for a long time you could have a double-check.

Reply:
I tried to refactor this but it is not easily possible. One of the reasons is that we cannot easily determine the working directory because we statically add `/.flink/appid` to the target path. I am actually ok with leaving this test.

Reply:
I would be in favor of simplifying `uploadAndRegisterFiles` in a separate commit.